File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/CTInAppBaseFullNativeFragment.java
package com.clevertap.android.sdk.inapp;

import android.content.Context;
import android.content.res.Configuration;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.LayerDrawable;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.RoundRectShape;
import android.os.Build;
import android.util.DisplayMetrics;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;

import com.clevertap.android.sdk.Logger;

public abstract class CTInAppBaseFullNativeFragment extends CTInAppBaseFullFragment {

    int getDPI() {
        WindowManager wm = (WindowManager) this.context.getSystemService(Context.WINDOW_SERVICE);
        if (wm == null) {
            return 0;
        }
        //Returns the dpi using Device Configuration API for API30 above
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            Configuration configuration = context.getResources().getConfiguration();
            return configuration.densityDpi;
        }else {
            DisplayMetrics dm = new DisplayMetrics();
            wm.getDefaultDisplay().getMetrics(dm);
            return dm.densityDpi;
        }
    }

    void setupInAppButton(Button inAppButton, final CTInAppNotificationButton inAppNotificationButton,
            final int buttonIndex) {
        if (inAppNotificationButton != null) {
            inAppButton.setVisibility(View.VISIBLE);
            inAppButton.setTag(buttonIndex);
            inAppButton.setText(inAppNotificationButton.getText());
            inAppButton.setTextColor(Color.parseColor(inAppNotificationButton.getTextColor()));
            inAppButton.setOnClickListener(new CTInAppNativeButtonClickListener());

            ShapeDrawable borderDrawable = null;
            ShapeDrawable shapeDrawable = null;

            if (!inAppNotificationButton.getBorderRadius().isEmpty()) {
                float value = Float.parseFloat(inAppNotificationButton.getBorderRadius()) * (480.0f / getDPI()) * 2;
                shapeDrawable = new ShapeDrawable(new RoundRectShape(new float[]{
                        value, value, value, value,
                        value, value, value, value
                }, null,
                        new float[]{0, 0, 0, 0, 0, 0, 0, 0}));
                shapeDrawable.getPaint().setColor(Color.parseColor(inAppNotificationButton.getBackgroundColor()));
                shapeDrawable.getPaint().setStyle(Paint.Style.FILL);
                shapeDrawable.getPaint().setAntiAlias(true);
                borderDrawable = new ShapeDrawable(new RoundRectShape(new float[]{
                        value, value, value, value,
                        value, value, value, value}, null,
                        new float[]{value, value, value, value,
                                value, value, value, value}));
            }

            if (!inAppNotificationButton.getBorderColor().isEmpty()) {
                if (borderDrawable != null) {
                    borderDrawable.getPaint().setColor(Color.parseColor(inAppNotificationButton.getBorderColor()));
                    borderDrawable.setPadding(1, 1, 1, 1);
                    borderDrawable.getPaint().setStyle(Paint.Style.FILL);
                }
            }

            if (shapeDrawable != null) {
                Drawable[] drawables = new Drawable[]{
                        borderDrawable,
                        shapeDrawable
                };

                LayerDrawable layerDrawable = new LayerDrawable(drawables);

                inAppButton.setBackground(layerDrawable);
            }
        } else {
            inAppButton.setVisibility(View.GONE);
        }
    }


}

File: clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/fragment/CTInAppBaseFullNativeFragment.kt
package com.clevertap.android.sdk.inapp.fragment

import android.content.Context
import android.graphics.Paint
import android.graphics.drawable.Drawable
import android.graphics.drawable.LayerDrawable
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.RoundRectShape
import android.os.Build
import android.util.DisplayMetrics
import android.view.View
import android.view.WindowManager
import android.widget.Button
import androidx.core.graphics.toColorInt
import com.clevertap.android.sdk.inapp.CTInAppNotificationButton

internal abstract class CTInAppBaseFullNativeFragment : CTInAppBaseFullFragment() {

    fun setupInAppButton(
        inAppButton: Button,
        inAppNotificationButton: CTInAppNotificationButton?,
        buttonIndex: Int
    ) {
        if (inAppNotificationButton != null) {
            inAppButton.visibility = View.VISIBLE
            inAppButton.tag = buttonIndex
            inAppButton.text = inAppNotificationButton.text
            inAppButton.setTextColor(inAppNotificationButton.textColor.toColorInt())
            inAppButton.setOnClickListener(CTInAppNativeButtonClickListener())

            var borderDrawable: ShapeDrawable? = null
            var shapeDrawable: ShapeDrawable? = null

            if (inAppNotificationButton.borderRadius.isNotEmpty()) {
                val buttonRadius = inAppNotificationButton.borderRadius.toFloatOrNull() ?: 0f
                val radiusDp: Float = buttonRadius * (480.0f / getDPI()) * 2
                shapeDrawable = ShapeDrawable(
                    RoundRectShape(
                        floatArrayOf(
                            radiusDp, radiusDp, radiusDp, radiusDp,
                            radiusDp, radiusDp, radiusDp, radiusDp
                        ), null,
                        floatArrayOf(0f, 0f, 0f, 0f, 0f, 0f, 0f, 0f)
                    )
                )
                shapeDrawable.paint.setColor(inAppNotificationButton.backgroundColor.toColorInt())
                shapeDrawable.paint.style = Paint.Style.FILL
                shapeDrawable.paint.isAntiAlias = true
                borderDrawable = ShapeDrawable(
                    RoundRectShape(
                        floatArrayOf(
                            radiusDp, radiusDp, radiusDp, radiusDp,
                            radiusDp, radiusDp, radiusDp, radiusDp
                        ), null,
                        floatArrayOf(
                            radiusDp, radiusDp, radiusDp, radiusDp,
                            radiusDp, radiusDp, radiusDp, radiusDp
                        )
                    )
                )
            }

            if (!inAppNotificationButton.borderColor.isEmpty()) {
                if (borderDrawable != null) {
                    borderDrawable.paint.setColor(inAppNotificationButton.borderColor.toColorInt())
                    borderDrawable.setPadding(1, 1, 1, 1)
                    borderDrawable.paint.style = Paint.Style.FILL
                }
            }

            if (shapeDrawable != null && borderDrawable != null) {
                val drawables: Array<Drawable> = arrayOf<Drawable>(
                    borderDrawable,
                    shapeDrawable
                )

                val layerDrawable = LayerDrawable(drawables)
                inAppButton.background = layerDrawable
            }
        } else {
            inAppButton.visibility = View.GONE
        }
    }

    private fun getDPI(): Int {
        val wm = requireContext().getSystemService(Context.WINDOW_SERVICE) as WindowManager?
        if (wm == null) {
            return DisplayMetrics.DENSITY_DEFAULT
        }
        //Returns the dpi using Device Configuration API for API30 above
        val density = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val configuration = requireContext().resources.configuration
            configuration.densityDpi
        } else {
            val dm = DisplayMetrics()
            wm.defaultDisplay.getMetrics(dm)
            dm.densityDpi
        }
        return if (density > 0) {
            density
        } else {
            DisplayMetrics.DENSITY_DEFAULT
        }
    }
}
